docs/specs/account-management-api.md

```diff
@@ -0,0 +1,861 @@
+```json
+{
+  "result": {
+    "identification_methods": [
+      {
+        "identification": "email",
+        "login_id": "user@example.com",
+        "claims": {
+          "email": "user@example.com"
+        },
+        "created_at": "2006-01-02T03:04:05Z",
+        "updated_at": "2006-01-02T03:04:05Z"
+      },
+      {
+        "identification": "username",
+        "login_id": "johndoe",
+        "claims": {
+          "preferred_username": "johndoe"
+        },
+        "created_at": "2006-01-02T03:04:05Z",
+        "updated_at": "2006-01-02T03:04:05Z"
+      },
+      {
+        "identification": "phone",
+        "login_id": "+85251000001",
+        "claims": {
+          "phone_number": "+85251000001"
+        },
+        "created_at": "2006-01-02T03:04:05Z",
+        "updated_at": "2006-01-02T03:04:05Z"
+      },
+      {
+        "identification": "oauth",
+        "provider_type": "google",
+        "provider_user_id": "USER_ID_AT_GOOGLE",
+        "alias": "google",
+        "claims": {
+          "email": "user@gmail.com"
+        },
+        "created_at": "2006-01-02T03:04:05Z",
+        "updated_at": "2006-01-02T03:04:05Z"
+      },
+      {
+        "identification": "passkey",
+        "credential_id": "hlNmaS6DQjV4voxP8SPJDzDG-j79nWL8r4OTgcPizi0"
+        "created_at": "2006-01-02T03:04:05Z",
+        "updated_at": "2006-01-02T03:04:05Z"
+      },
+      {
+        "identification": "biometric",
+        "key_id": "KEY_ID",
+        "display_name": "iPhone 12 mini",
+        "created_at": "2006-01-02T03:04:05Z",
+        "updated_at": "2006-01-02T03:04:05Z"
+      }
+    ]
+  }
+}
+```
+
+### Add a new email, with verification
+
+`POST /api/v1/account/identification`
+
+Request
+
+```json
+{
+  "identification": "email",
+  "login_id": "user@example.com"
+}
+```
+
+Response
+
+If verification is not required, then the email is added immediately.
+
+```json
+{
+  "result": {
+    "identification_method": {
+      "identification": "email",
+      "login_id": "user@example.com",
+      "claims": {
+        "email": "user@example.com"
+      },
+      "created_at": "2006-01-02T03:04:05Z",
+      "updated_at": "2006-01-02T03:04:05Z"
+    }
+  }
+}
+```
+
+If verification is required, you need to perform verification.
+
+```json
+{
+  "result": {
+    "verification": {
+      "token": "verificationtoken_blahblahblah",
+      "channel": "email",
+      "otp_form": "code",
+      "code_length": 6,
+      "can_resend_at": "2023-09-21T00:00:00+08:00",
+      "can_check": false,
+      "failed_attempt_rate_limit_exceeded": false
+    }
+  }
+}
+```
+
+Use [Verify OTP](#verify-otp) and [Resend OTP](#resend-otp) to continue.
+
+If the email is already taken by another account, then you will receive the following error.
+
+```json
+{
+  "error": {
+    "name": "Invalid",
+    "reason": "InvariantViolated",
+    "message": "identity already exists",
+    "code": 400,
+    "info": {
+      "cause": {
+        "kind": "DuplicatedIdentity"
```

Comment:
The error object has the following fields. It has been used very long ago, and it is not invented by this spec.

- `name`: It is actually the category of the error.
- `code`: The HTTP status code of the error.
- `reason`: This is the main field we distinguish between different errors
- `info`: Extra information for a particular `reason`

The problem here is that we overload the reason `InvariantViolated` to represent some important errors. Opened another issue for this https://github.com/authgear/authgear-server/issues/4422

Reply:
@louischan-oursky For #4422 can we actually try to improve our error object? 

1. Besides this API and Authflow, where is it being used?
2. I think instead of:

```json
"error": {
    "name": "Invalid",
    "reason": "InvariantViolated",
    "message": "identity already exists",
    "code": 400,
    "info": {
      "cause": {
        "kind": "DuplicatedIdentity"
      }
    }
}
```

We shall change to:
```
"error": {
  "type": "TypeOfError"
  "code": "Short String for the Error"
  "message": "Human readable message"
  "info": { }
}
```

The main changes from current design are:
- Use `type` instead of `name` for category of error. Name is very confusing
- HTTP status code should go back to HTTP response header
- `code` and `message` are the actual error
- We shall study the APIs used this error scheme, and try to come up with a standardized object for `info`, otherwise it is not usable at all (or else we need to write a spec to indicate what error will have which `info`, but it is not as good as a standardized object.

Let me copy this above message to #4422 as well.

Reply:
Continue the discussion in https://github.com/authgear/authgear-server/issues/4422